models/project.go

```diff
@@ -0,0 +1,296 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"errors"
+
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/timeutil"
+	"code.gitea.io/gitea/modules/util"
+)
+
+type (
+	// ProjectsConfig is used to identify the type of board that is being created
+	ProjectsConfig struct {
+		BoardType   ProjectBoardType
+		Translation string
+	}
+
+	// ProjectType is used to identify the type of project in question and ownership
+	ProjectType uint8
+)
+
+const (
+	// IndividualType is a type of project board that is owned by an individual
+	IndividualType ProjectType = iota + 1
+
+	// RepositoryType is a project that is tied to a repository
+	RepositoryType
+
+	// OrganizationType is a project that is tied to an organisation
+	OrganizationType
+)
+
+// Project represents a project board
+type Project struct {
+	ID          int64  `xorm:"pk autoincr"`
+	Title       string `xorm:"INDEX NOT NULL"`
+	Description string `xorm:"TEXT"`
+	RepoID      int64  `xorm:"NOT NULL"`
```

Comment:
How does this work out if the project is an `IndividualType`? The description of the types would seem to imply that RepoID could be NULL, but this constraint makes me think otherwise.

Reply:
this field will be used for the feature (not here jet) "Projects per User" & "Organisation Projects"

Reply:
Ok, those aren't implemented yet... it still feels like the design might be a bit off, though-

1. RepoID is `NOT NULL`, so this will later require a schema change as soon as any other type does get implemented
2. A Repository's projects are unconditionally incremented when a project is created, rather than only for the currently supported project type
3. Similar assumption in `ChangeProjectStatus`
4. On the basis of the above, it seems like it might be a good idea to call the other types invalid for now in `IsProjectTypeValid` until they are implemented with the necessary accounting, to prevent foot-shooting